src/verifiers/sev-snp/crypto.c

```diff
@@ -0,0 +1,222 @@
+/* Copyright (c) 2022 Intel Corporation
+ * Copyright (c) 2020-2022 Alibaba Cloud
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <openssl/sha.h>
+#include <openssl/ts.h>
+#include <rats-tls/log.h>
+#include "crypto.h"
+
+typedef enum __attribute__((mode(QI))) SHA_TYPE {
+	SHA_TYPE_256 = 0,
+	SHA_TYPE_384 = 1,
+} SHA_TYPE;
+
+#define BITS_PER_BYTE 8
+
+/* The caller of digest_sha() must ensure the @digest_len is
+ * correctly corresponding to @sha_type
+ */
+static bool digest_sha(const void *msg, size_t msg_len, uint8_t *digest, size_t digest_len,
+		       SHA_TYPE sha_type)
+{
+	switch (sha_type) {
+	case SHA_TYPE_256:
+		if (digest_len != SHA256_DIGEST_LENGTH)
+			return false;
```

Comment:
You have already noted the caller must ensure the @digest_len is correctly corresponding to @sha_type. If it really happens, it means the caller doesn't follow the principle. So you could just add `assert()` here rather than a sanity check.